Add available space to devicemapper status output

petervo opened the pull request:
PR for #10473.

petervo committed:
Add available space to devicemapper status output

Signed-off-by: Peter Volpe <petervo@redhat.com>

jfrazelle commented:
Can you please sign your commits following these rules:

https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work

The easiest way to do this is to amend the last commit:

`git commit --amend -s --no-edit && git push -f`

jfrazelle commented:
ping @vbatts

vbatts commented on daemon/graphdriver/devmapper/deviceset.go:
```diff
@@ -1595,12 +1618,28 @@ func (devices *DeviceSet) Status() *Status {
 
 		status.SectorSize = blockSizeInSectors * 512
+
+		if check, _ := devices.isRealFile(devices.dataLoopFile); check {
```
This check of whether it is a real loop file is not needed now. If `devices.dataLoopFile != ""` then there is a loop file.

petervo replied:
That was to account for if the loopfile is a link to a device as described here (http://jpetazzo.github.io/2014/01/29/docker-device-mapper-resize/) see "We need a faster pool" section.

vbatts commented:
@petervo this looks to primarily address loopfile, but will it be same behavior if there is _no_ loop file, and a direct block device mounted for `devices.dataDevice` and `devices.metadataDevice`? Perhaps the code need only a bit of refactor such that the statfs only accounts for the mounted device, regardless of loop or not.

petervo commented:
@vbatts I'm not a devicemapper expert, so feel free to tell me i'm wrong with any of my statements, this is to the best of my limited knowlege.

The goal here was to have a way to get the actual available space when the device is backed by a sparse file and do the naive (total - used) everywhere else.  My understanding is that using a sparse file is only possible with a loop device, if we wanted we could support sparse files that were not created by docker by using stat to check if the device is a loop device, getting the target and checking that is a real file before proceeding with the statfs call.

We can't call statfs on block devices themselves so when the device is not a loop device we don't know that the device isn't overstating it's capacity ei a lvm thin volume. So I'm not sure what the solution is there but that is a custom configuration so presumably they had a good reason for doing that and another way to monitor capacity, and not something docker just did out of the box.

vbatts commented on daemon/graphdriver/devmapper/README.md:
```diff
@@ -60,8 +62,10 @@ status information about the driver.
  *  `Metadata file` blockdevice file used for the devicemapper metadata
  *  `Data Space Used` tells how much of `Data file` is currently used
  *  `Data Space Total` tells max size the `Data file`
+ *  `Data Space Available` tells how much free space there is in the `Data file`
```
Perhaps, lets add a parenthetical wording regarding the adjustment made for underlying filesystem space, when the device is a loop file

vbatts commented:
Right, so this is less about the sparseness of this, but the consideration of the underlying file system, when the loop file is used. I like that a lot. 

Overall, LGTM besides the nit of documentation wording. Since there will be scenarios that the number don't balance, and it should be clear to the users.

petervo committed:
Detail available space behavior on loop devices

Signed-off-by: Peter Volpe <petervo@redhat.com>

petervo commented:
I added a line to the documentation.

Based off my response to your earlier comments i also added a commit to support loop devices that docker didn't setup. I'm not sure if that's a requirement for this or not. It does rely on /sys/dev being present which is kernel 2.6+. If we need to support earlier than that we'll need to use the fallback that is in losetup. let me know if I need to add that. Or if you prefer we can leave that commit out and just only do the statfs check when docker already knows it's dealing with a loop device.

vbatts commented:
@petervo let's not do that logic to determine the attached loopfile. That is why I recently rearranged a bit to have a field in DeviceSet for the loop file _if_ it is a loop file https://github.com/docker/docker/blob/master/daemon/graphdriver/devmapper/deviceset.go#L93
So, if the `dataLoopFile` and `metadataLoopFile` are not `== ""` then just use that value.

petervo commented:
Yeah that makes sense was a edge case, I removed that commit. Are there other changes needed?

vbatts commented:
no other changes needed. You've also fixed my doc nit too.

@unclejack or @tiborvass  could you look this once over before i merge it?

unclejack commented:
@vbatts I'm going to try it out now and review.

unclejack commented:
LGTM

vbatts commented:
LGTM
